From 2b5961558c14d5cfa68cf99b8d1d51f8fd9ba734 Mon Sep 17 00:00:00 2001
From: John Thomson <git@johnthomson.fastmail.com.au>
Date: Thu, 27 Oct 2022 20:56:05 +1000
Subject: [PATCH 1/6] ramips: mt7621: mikrotik: initram use lzma loader

Use OpenWrt's lzma loader for initramfs image
Due to mips kernel self extractor inability to relocate,
and RouterBoot on 760igs refusing to netboot an ELF with:
Entry point address:               0x80b91000
or greater: 0x80b81000 and below okay.

Note that this is only a bootloader ELF entry point limitation,
a 50MiB Linux + initramfs vmlinux will boot fine on 760igs.

With this lzma loader, netboot will occasionally freeze after
decompressing the kernel, but before starting vmlinux. It was considered
that an occasional netboot freeze was simpler to workaround than an
image that refuses to boot unless minimized.

Signed-off-by: John Thomson <git@johnthomson.fastmail.com.au>
---
 target/linux/ramips/image/mt7621.mk | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/target/linux/ramips/image/mt7621.mk b/target/linux/ramips/image/mt7621.mk
index 9654e237a8af22..63a67ce8fbcda7 100644
--- a/target/linux/ramips/image/mt7621.mk
+++ b/target/linux/ramips/image/mt7621.mk
@@ -2000,6 +2000,9 @@ define Device/MikroTik
   DEVICE_PACKAGES := kmod-usb3 -uboot-envtools
   KERNEL_NAME := vmlinuz
   KERNEL := kernel-bin | append-dtb-elf
+  LOADER_TYPE := elf
+  KERNEL_INITRAMFS_NAME := vmlinux-initramfs
+  KERNEL_INITRAMFS := $(KERNEL_DTB) | loader-kernel
   IMAGE/sysupgrade.bin := append-kernel | yaffs-filesystem -L | \
 	pad-to $$$$(BLOCKSIZE) | append-rootfs | pad-rootfs | check-size | \
 	append-metadata

From 831c93f0aa04617902dc6ce229e945c2af3591d6 Mon Sep 17 00:00:00 2001
From: John Thomson <git@johnthomson.fastmail.com.au>
Date: Sun, 22 Jan 2023 21:34:43 +1000
Subject: [PATCH 2/6] build: add Mikrotik NOR RouterBOOT v7 image build

Signed-off-by: John Thomson <git@johnthomson.fastmail.com.au>
---
 include/image-commands.mk |  5 +++++
 tools/Makefile            |  1 +
 tools/npkpy/Makefile      | 33 +++++++++++++++++++++++++++++++++
 3 files changed, 39 insertions(+)
 create mode 100644 tools/npkpy/Makefile

diff --git a/include/image-commands.mk b/include/image-commands.mk
index 2568fb3ade5f3d..dbb032b9a4f93a 100644
--- a/include/image-commands.mk
+++ b/include/image-commands.mk
@@ -470,6 +470,11 @@ define Build/kernel-bin
 	cp $< $@
 endef
 
+define Build/kernel-pack-npk
+	sh -c '. $(STAGING_DIR_HOST)/share/npkpy-venv/bin/activate; python $(STAGING_DIR_HOST)/share/npkpy-venv/bin/pack_npk_kernel.py --kernel $@ --output $@.npk'
+	mv $@.npk $@
+endef
+
 define Build/linksys-image
 	let \
 		size="$$(stat -c%s $@)" \
diff --git a/tools/Makefile b/tools/Makefile
index b16c5d9c5b1e6c..ea494911a1d9d3 100644
--- a/tools/Makefile
+++ b/tools/Makefile
@@ -86,6 +86,7 @@ tools-$(if $(CONFIG_BUILD_ALL_HOST_TOOLS)$(CONFIG_USES_MINOR),y) += yafut
 tools-$(if $(CONFIG_BUILD_ALL_HOST_TOOLS)$(CONFIG_USE_SPARSE),y) += sparse
 tools-$(if $(CONFIG_BUILD_ALL_HOST_TOOLS)$(CONFIG_USE_LLVM_BUILD),y) += llvm-bpf
 tools-$(if $(CONFIG_BUILD_ALL_HOST_TOOLS)$(CONFIG_USE_MOLD),y) += mold
+tools-$(if $(CONFIG_BUILD_ALL_HOST_TOOLS)$(CONFIG_USES_MINOR),y) += npkpy
 
 # builddir dependencies
 $(curdir)/autoconf/compile := $(curdir)/m4/compile
diff --git a/tools/npkpy/Makefile b/tools/npkpy/Makefile
new file mode 100644
index 00000000000000..c3bd7921016c7a
--- /dev/null
+++ b/tools/npkpy/Makefile
@@ -0,0 +1,33 @@
+#
+# Copyright (C) 2023 OpenWrt.org
+#
+# This is free software, licensed under the GNU General Public License v2.
+# See /LICENSE for more information.
+#
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=npkpy
+PKG_RELEASE:=1
+
+PKG_SOURCE_PROTO:=git
+PKG_SOURCE_URL:=https://github.com/john-tho/npkpy.git
+PKG_SOURCE_DATE:=2023-01-22
+PKG_SOURCE_VERSION:=771dddc5aaaff95a9b4160d6867152cfa1a9255f
+PKG_MIRROR_HASH:=4c0596218e8c1f78b38c0583c18a574da18ff4c000ecdd5fbf50c62a62a1fb03
+
+include $(INCLUDE_DIR)/host-build.mk
+
+define Host/Compile
+	python -m venv $(STAGING_DIR_HOST)/share/npkpy-venv
+	sh -c '. $(STAGING_DIR_HOST)/share/npkpy-venv/bin/activate; pip install --editable $(HOST_BUILD_DIR)'
+endef
+
+define Host/Install
+	$(CP) $(HOST_BUILD_DIR)/tools/demo_pack_kernel/poc_pack_kernel.py $(STAGING_DIR_HOST)/share/npkpy-venv/bin/pack_npk_kernel.py
+endef
+
+define Host/Clean
+	rm -rf $(STAGING_DIR_HOST)/share/npkpy-venv
+endef
+
+$(eval $(call HostBuild))

From 71d80889ba3c2c6d2a380f11e6268a9f98e5ee6b Mon Sep 17 00:00:00 2001
From: John Thomson <git@johnthomson.fastmail.com.au>
Date: Thu, 16 May 2024 10:16:53 +1000
Subject: [PATCH 3/6] image: adapt mikrotik yafut to allow switch to bootimage
 YAFFS path

Adjust the YAFFS file path written depending on the sysupgrade
filename.
Default to kernel (for ELF), switch to bootimage (for NPK) if image name
has v7.

Signed-off-by: John Thomson <git@johnthomson.fastmail.com.au>
---
 include/image-commands.mk | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/image-commands.mk b/include/image-commands.mk
index dbb032b9a4f93a..fa2bce1038e747 100644
--- a/include/image-commands.mk
+++ b/include/image-commands.mk
@@ -460,7 +460,7 @@ define Build/yaffs-filesystem
 		filesystem_size="filesystem_blocks * 64 * 1024" \
 		filesystem_size_with_reserve="(filesystem_blocks + 2) * 64 * 1024"; \
 		head -c $$filesystem_size_with_reserve /dev/zero | tr "\000" "\377" > $@.img \
-		&& yafut -d $@.img -w -i $@ -o kernel -C 1040 -B 64k -E -P -S $(1) \
+		&& yafut -d $@.img -w -i $@ -o $(if $(findstring v7,$@),bootimage,kernel) -C 1040 -B 64k -E -P -S $(1) \
 		&& truncate -s $$filesystem_size $@.img \
 		&& mv $@.img $@
 endef

From bbab5a78cc0779777ebc39eb2440dd629b66b373 Mon Sep 17 00:00:00 2001
From: John Thomson <git@johnthomson.fastmail.com.au>
Date: Mon, 23 Jan 2023 07:16:04 +1000
Subject: [PATCH 4/6] ipq40xx: mikrotik: generate a RouterBOOT v7 NOR
 compatible sysupgrade

RouterBOOT v7 for NOR devices does not support the historic yaffs
"kernel" ELF boot method.
Generate a compatible kernel

Signed-off-by: John Thomson <git@johnthomson.fastmail.com.au>
---
 target/linux/ipq40xx/image/mikrotik.mk | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/target/linux/ipq40xx/image/mikrotik.mk b/target/linux/ipq40xx/image/mikrotik.mk
index 8c2c6fa08edfe9..5d6e3c9808a4f6 100644
--- a/target/linux/ipq40xx/image/mikrotik.mk
+++ b/target/linux/ipq40xx/image/mikrotik.mk
@@ -5,9 +5,13 @@ define Device/mikrotik_nor
 	KERNEL_NAME := vmlinux
 	KERNEL := kernel-bin | append-dtb-elf
 	IMAGES = sysupgrade.bin
+	IMAGES += sysupgrade-v7.bin
 	IMAGE/sysupgrade.bin := append-kernel | yaffs-filesystem -L | \
 		pad-to $$$$(BLOCKSIZE) | append-rootfs | pad-rootfs | \
 		check-size | append-metadata
+	IMAGE/sysupgrade-v7.bin := append-kernel | kernel-pack-npk | \
+		yaffs-filesystem -L | pad-to $$$$(BLOCKSIZE) | \
+		append-rootfs | pad-rootfs | check-size | append-metadata
 endef
 
 define Device/mikrotik_nand

From 97c46fd224269381cfacd231919724669c58c3d9 Mon Sep 17 00:00:00 2001
From: John Thomson <git@johnthomson.fastmail.com.au>
Date: Mon, 23 Jan 2023 07:16:04 +1000
Subject: [PATCH 5/6] ramips: mikrotik: generate a RouterBOOT v7 NOR compatible
 sysupgrade

RouterBOOT v7 for NOR devices does not support the historic yaffs
"kernel" ELF boot method.
Generate a compatible kernel

Signed-off-by: John Thomson <git@johnthomson.fastmail.com.au>
---
 target/linux/ramips/image/mt7621.mk | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/target/linux/ramips/image/mt7621.mk b/target/linux/ramips/image/mt7621.mk
index 63a67ce8fbcda7..32b96130b7bfc7 100644
--- a/target/linux/ramips/image/mt7621.mk
+++ b/target/linux/ramips/image/mt7621.mk
@@ -2006,6 +2006,10 @@ define Device/MikroTik
   IMAGE/sysupgrade.bin := append-kernel | yaffs-filesystem -L | \
 	pad-to $$$$(BLOCKSIZE) | append-rootfs | pad-rootfs | check-size | \
 	append-metadata
+  IMAGES += sysupgrade-v7.bin
+  IMAGE/sysupgrade-v7.bin := append-kernel | kernel-pack-npk | \
+	  yaffs-filesystem -L | pad-to $$$$(BLOCKSIZE) | \
+	  append-rootfs | pad-rootfs | check-size | append-metadata
 endef
 
 define Device/mikrotik_ltap-2hnd

From ab83ff59d7a00b7a992600c5743421fafe3987d7 Mon Sep 17 00:00:00 2001
From: John Thomson <git@johnthomson.fastmail.com.au>
Date: Thu, 27 Oct 2022 20:56:05 +1000
Subject: [PATCH 6/6] ath79: mikrotik: generate a RouterBOOT v7 NOR compatible
 sysupgrade

RouterBOOT v7 for NOR devices does not support the historic yaffs
"kernel" ELF boot method.
Generate a compatible kernel

Signed-off-by: John Thomson <git@johnthomson.fastmail.com.au>
---
 target/linux/ath79/image/common-mikrotik.mk | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/target/linux/ath79/image/common-mikrotik.mk b/target/linux/ath79/image/common-mikrotik.mk
index 94c29d8cb6ab04..1dfd515d25e51e 100644
--- a/target/linux/ath79/image/common-mikrotik.mk
+++ b/target/linux/ath79/image/common-mikrotik.mk
@@ -13,6 +13,10 @@ define Device/mikrotik_nor
   IMAGE/sysupgrade.bin := append-kernel | yaffs-filesystem -M | \
 	pad-to $$$$(BLOCKSIZE) | append-rootfs | pad-rootfs | \
 	check-size | append-metadata
+  IMAGES += sysupgrade-v7.bin
+  IMAGE/sysupgrade-v7.bin := append-kernel | kernel-pack-npk | \
+	  yaffs-filesystem -M | pad-to $$$$(BLOCKSIZE) | \
+	  append-rootfs | pad-rootfs | check-size | append-metadata
 endef
 
 define Device/mikrotik_nand
